src/classes/HiFiCommunicator.ts

```diff
@@ -159,13 +300,21 @@ export class HiFiCommunicator {
         if (onUsersDisconnected) {
             this.onUsersDisconnected = onUsersDisconnected;
         }
+        if (onConnectionStateChanged) {
+            this.onConnectionStateChanged = onConnectionStateChanged;
+        }
+        this._connectionRetryAndTimeoutConfig = HiFiConstants.DEFAULT_CONNECTION_RETRY_AND_TIMEOUT;
+        if (connectionRetryAndTimeoutConfig) {
+            Object.assign(this._connectionRetryAndTimeoutConfig, connectionRetryAndTimeoutConfig);
+        }
 
         this._mixerSession = new HiFiMixerSession({
             "userDataStreamingScope": userDataStreamingScope,
             "onUserDataUpdated": (data: Array<ReceivedHiFiAudioAPIData>) => { this._handleUserDataUpdates(data); },
             "onUsersDisconnected": (data: Array<ReceivedHiFiAudioAPIData>) => { this._onUsersDisconnected(data); },
-            "onConnectionStateChanged": onConnectionStateChanged,
-            "onMuteChanged": onMuteChanged
+            "onConnectionStateChanged": (state: HiFiConnectionStates, message: ConnectionAttemptResult) => { this._manageConnection(state, message); },
+            "onMuteChanged": onMuteChanged,
+            "getUserFacingConnectionState": this.getConnectionState.bind(this)
```

Comment:
I didn't follow what this is a about (or for that matter, what getConnectionState is for as a function rather than just letting the state property be public, as it seems to only be used here, but I guess that's a separate question).

It looks like maybe the only reason for this is to decide whether to call the  onConnectionStateChanged that we pass in here, but I don't understand why not just alway call it and let the manager decide what to do about it. After all, the manager already has access to that info.

Reply:
Ah -- the main reason for this is that:
1) The connection state is now tracked on the HiFiCommunicator, but
2) The diagnostics code only knows about the HiFiMixerSession, not the HiFiCommunicator, and
3) The HiFiMixerSession doesn't have a reference to the HiFiCommunicator that's using it, so
4) In order for the diagnostics code to be able to find out what the current "user facing" state is, it needs to be able to get that from the HiFiMixerSession (since that's all it knows about), so
5) The HiFiMixerSession needs to be able to get it from the HiFiCommunicator, so
6) This is a way for the HiFiMixerSession to get the connection state that the HiFiCommunicator knows about without having to pass a reference to the ENTIRE HiFiCommunicator into the HiFiMixerSession.

Reply:
If this is here JUST for the diagnostics code....  I'd rather have the session/state/management code be clean than to have the diagnostics code be clean. (I could create/expose some hooks in the MixerSession, and poke those hooks from some temporary code in the Communicator.)

Reply:
Yep, it's there just for the diagnostics code. Both because the diagnostics code needs to get the "current user-visible state" and also because HiFiMixerSession's `_onConnectionStateChange` method needs to be able to check to see if the state being passed is different than the current state (so that it can decide when/if to fire the diagnostics).

I mean TBH I actually think this is a reasonably elegant way of passing the information down (I was proud of myself for thinking of it 😄 ) , but I'm also fine with your exposing temporary hooks to store a copy of the info in the MixerSession instead. Do you want to flag this and rework it post-merge, or add a commit to this PR? (Or I can give it a shot if you'd like.)

Reply:
I think we can clean it up later. The main thing is to understand what the design is.

Reply:
Sorry for having made this work more complicated!

Reply:
Not at all! I'll add a comment in that general area with a TODO about cleaning it up later.

Reply:
(Comment added to the HiFiMixerSession's parameter explanation. Also added comments for other HiFiMixerSession parameters that were previously uncommented, while I was in there.)